File: examples/go/token-nft/NFT_test.go
package contract

import (
	"testing"
	runar "github.com/icellan/runar/packages/runar-go"
)

func newNFT(owner runar.PubKey) *SimpleNFT {
	return &SimpleNFT{
		Owner:    owner,
		TokenId:  runar.ByteString("unique-nft-001"),
		Metadata: runar.ByteString("ipfs://QmTest"),
	}
}

func TestSimpleNFT_Transfer(t *testing.T) {
	alice := runar.PubKey("alice_pubkey_33bytes_placeholder!")
	bob := runar.PubKey("bob___pubkey_33bytes_placeholder!")
	c := newNFT(alice)
	c.Transfer(runar.MockSig(), bob, 1000)
	if len(c.Outputs()) != 1 {
		t.Fatalf("expected 1 output, got %d", len(c.Outputs()))
	}
}

func TestSimpleNFT_Burn(t *testing.T) {
	alice := runar.PubKey("alice_pubkey_33bytes_placeholder!")
	c := newNFT(alice)
	c.Burn(runar.MockSig())
	if len(c.Outputs()) != 0 {
		t.Errorf("expected 0 outputs after burn, got %d", len(c.Outputs()))
	}
}

func TestSimpleNFT_Compile(t *testing.T) {
	if err := runar.CompileCheck("NFTExample.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

File: examples/rust/token-nft/NFT_test.rs
// Contract logic tests for SimpleNFT.
//
// The contract struct is defined inline because add_output output-tracking
// requires test infrastructure not part of the TSOP contract.

use tsop::prelude::*;

#[derive(Clone)]
struct NftOutput { satoshis: Bigint, owner: PubKey }

struct SimpleNFT {
    owner: PubKey,
    token_id: ByteString,
    metadata: ByteString,
    outputs: Vec<NftOutput>,
}

impl SimpleNFT {
    fn add_output(&mut self, satoshis: Bigint, new_owner: PubKey) {
        self.outputs.push(NftOutput { satoshis, owner: new_owner });
    }

    fn transfer(&mut self, sig: &Sig, new_owner: PubKey, output_satoshis: Bigint) {
        assert!(check_sig(sig, &self.owner));
        self.add_output(output_satoshis, new_owner);
    }

    fn burn(&self, sig: &Sig) {
        assert!(check_sig(sig, &self.owner));
    }
}

fn alice() -> PubKey { b"alice_pubkey_33bytes_placeholder!".to_vec() }
fn bob() -> PubKey { b"bob___pubkey_33bytes_placeholder!".to_vec() }
fn charlie() -> PubKey { b"charlie_pk_33bytes_placeholder__".to_vec() }

fn new_nft(owner: PubKey) -> SimpleNFT {
    SimpleNFT {
        owner,
        token_id: b"unique-nft-001".to_vec(),
        metadata: b"ipfs://QmTest".to_vec(),
        outputs: vec![],
    }
}

#[test]
fn test_transfer() {
    let mut c = new_nft(alice());
    c.transfer(&mock_sig(), bob(), 1000);
    assert_eq!(c.outputs.len(), 1);
    assert_eq!(c.outputs[0].owner, bob());
}

#[test]
fn test_transfer_chain() {
    let mut c = new_nft(alice());
    c.transfer(&mock_sig(), bob(), 1000);
    c.owner = bob();
    c.outputs.clear();
    c.transfer(&mock_sig(), charlie(), 1000);
    assert_eq!(c.outputs[0].owner, charlie());
}

#[test]
fn test_burn() {
    let c = new_nft(alice());
    c.burn(&mock_sig());
    assert_eq!(c.outputs.len(), 0);
}

#[test]
fn test_compile() {
    tsop::compile_check(include_str!("NFTExample.tsop.rs"), "NFTExample.tsop.rs").unwrap();
}
